x/auth/vesting/internal/types/vesting_account.go

```diff
@@ -0,0 +1,557 @@
+package types
+
+import (
+	"errors"
+	"time"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/x/auth"
+	authexported "github.com/cosmos/cosmos-sdk/x/auth/exported"
+	authtypes "github.com/cosmos/cosmos-sdk/x/auth/types"
+	vestexported "github.com/cosmos/cosmos-sdk/x/auth/vesting/exported"
+	"gopkg.in/yaml.v2"
+)
+
+// Assert BaseVestingAccount implements authexported.Account at compile-time
+var _ authexported.Account = (*BaseVestingAccount)(nil)
+
+// Assert vesting accounts implement vestexported.VestingAccount at compile-time
+var _ vestexported.VestingAccount = (*ContinuousVestingAccount)(nil)
+var _ vestexported.VestingAccount = (*PeriodicVestingAccount)(nil)
+var _ vestexported.VestingAccount = (*DelayedVestingAccount)(nil)
+
+// Register the vesting account types on the auth module codec
+func init() {
+	authtypes.RegisterAccountTypeCodec(&BaseVestingAccount{}, "cosmos-sdk/BaseVestingAccount")
+	authtypes.RegisterAccountTypeCodec(&ContinuousVestingAccount{}, "cosmos-sdk/ContinuousVestingAccount")
+	authtypes.RegisterAccountTypeCodec(&DelayedVestingAccount{}, "cosmos-sdk/DelayedVestingAccount")
+	authtypes.RegisterAccountTypeCodec(&PeriodicVestingAccount{}, "cosmos-sdk/PeriodicVestingAccount")
+}
+
+// BaseVestingAccount implements the VestingAccount interface. It contains all
+// the necessary fields needed for any vesting account implementation.
+type BaseVestingAccount struct {
+	*auth.BaseAccount
+	OriginalVesting  sdk.Coins      `json:"original_vesting" yaml:"original_vesting"`   // coins in account upon initialization
+	DelegatedFree    sdk.Coins      `json:"delegated_free" yaml:"delegated_free"`       // coins that are vested and delegated
+	DelegatedVesting sdk.Coins      `json:"delegated_vesting" yaml:"delegated_vesting"` // coins that vesting and delegated
+	EndTime          int64          `json:"end_time" yaml:"end_time"`                   // when the coins become unlocked
+	StartTime        int64          `json:"start_time,omitempty" yaml:"start_time,omitempty"`
+	VestingPeriods   VestingPeriods `json:"vesting_periods,omitempty" yaml:"vesting_periods,omitempty"`
+}
+
+// NewBaseVestingAccount creates a new BaseVestingAccount object
+func NewBaseVestingAccount(baseAccount *auth.BaseAccount, originalVesting sdk.Coins,
+	endTime int64) *BaseVestingAccount {
+
+	return &BaseVestingAccount{
+		BaseAccount:      baseAccount,
+		OriginalVesting:  originalVesting,
+		DelegatedFree:    sdk.NewCoins(),
+		DelegatedVesting: sdk.NewCoins(),
+		EndTime:          endTime,
+	}
+}
+
+// spendableCoins returns all the spendable coins for a vesting account given a
+// set of vesting coins.
+//
+// CONTRACT: The account's coins, delegated vesting coins, vestingCoins must be
+// sorted.
+func (bva BaseVestingAccount) spendableCoins(vestingCoins sdk.Coins) sdk.Coins {
```

Comment:
I ran in to an issue when writing a custom vesting account that makes me think `spendableCoins` and `trackDelegation` shouldn't be private methods on `BaseVestingAccount` because it breaks composability. 

The basic issue is that I have a new Vesting Account, which implements `GetVestingCoins` like so:
```
func (vva ValidatorVestingAccount) GetVestingCoins(blockTime time.Time) sdk.Coins {
	return vva.OriginalVesting.Sub(vva.GetVestedCoins(blockTime)).Sub(vva.GetFailedVestedCoins(blockTime))
}
```

Since I changed `GetVestingCoins`, I also need to change functions in methods that are inherited from `ContinuousVestingAccount` that also call `GetVestingCoins`, namely `SpendableCoins`. However, `SpendableCoins` is defined as follows:

```
// SpendableCoins returns the total number of spendable coins per denom for a
// continuous vesting account.
func (cva ContinuousVestingAccount) SpendableCoins(blockTime time.Time) sdk.Coins {
	return cva.spendableCoins(cva.GetVestingCoins(blockTime))
}
```

Because I'm defining `SpendableCoins` from outside the `vesting` package, I don't have access to `spendableCoins`, which thus means I have copy/paste `spendableCoins` and lose composability. 

@alexanderbez @fedekunze Do you agree with my view of the problem and is it alright if I make these methods public?

Reply:
@karzak I see no problem in exposing `spendableCoins` and `trackDelegation` for the `BaseVestingAccount` type -- totally OK.